test/byContract/IexecPoco/IexecPoco2-contribute-and-finalize.test.ts

```diff
@@ -0,0 +1,515 @@
+            const { dealId, taskIndex, taskId } = await iexecWrapper.signAndMatchOrders(
+                ...buildOrders({
+                    assets: ordersAssets,
+                    requester: requester.address,
+                    prices: ordersPrices,
+                    volume,
+                    trust: 3,
+                    tag: standardDealTag,
+                }).toArray(),
+            );
+            await iexecPoco.initialize(dealId, taskIndex).then((tx) => tx.wait());
+            // Task active, before deadline, but bad trust.
+            await expect(
+                iexecPoco
+                    .connect(worker)
+                    .contributeAndFinalize(
+                        taskId,
+                        resultDigest,
+                        results,
+                        noCallbackData,
+                        emptyEnclaveAddress,
+                        emptyEnclaveSignature,
+                        await buildAndSignContributionAuthorizationMessage(
+                            worker.address,
+                            taskId,
+                            emptyEnclaveAddress,
+                            scheduler,
+                        ),
+                    ),
+            ).to.be.revertedWithoutReason(); // require#4
+        });
+
+        it('Should not contributeAndFinalize when callback data is missing', async () => {
+            // Create deal and task.
+            const { dealId, taskIndex, taskId } = await iexecWrapper.signAndMatchOrders(
+                ...buildOrders({
+                    assets: ordersAssets,
+                    requester: requester.address,
+                    prices: ordersPrices,
+                    volume,
+                    trust,
+                    tag: standardDealTag,
+                    callback: ethers.Wallet.createRandom().address, // Using callback
+                }).toArray(),
+            );
+            await iexecPoco.initialize(dealId, taskIndex).then((tx) => tx.wait());
+            // Task active, before deadline, good trust, but missing callback data.
+            const { callbackResultDigest: resultsCallbackDigest } =
+                buildResultCallbackAndDigest(123);
+            await expect(
+                iexecPoco
+                    .connect(worker)
+                    .contributeAndFinalize(
+                        taskId,
+                        resultsCallbackDigest,
+                        noResultsData,
+                        noCallbackData,
+                        emptyEnclaveAddress,
+                        emptyEnclaveSignature,
+                        await buildAndSignContributionAuthorizationMessage(
+                            worker.address,
+                            taskId,
+                            emptyEnclaveAddress,
+                            scheduler,
+                        ),
+                    ),
+            ).to.be.revertedWithoutReason(); // require#5
+        });
+
```

Comment:
I think we are missing this one:
```suggestion

   it('Should not contributeAndFinalize when callback data is not hash of digest, async () => { .. });

```

Reply:
I have thought about it but actually in this case "missing callback data" is the same as "bad callback" because when `bytes calldata _resultsCallback` is empty, its hash is computed (a hash of empty bytes) and renders a non matching hash which is the same as sending bad data.
I can however change the name of the test to "bad or missing callback data" and send a bad non empty callback data payload. I think there is no point in duplicating the test. Would that be ok for you?

Reply:
Yes, we can keep 1 single test, and you are right, the test is about verifying that there is a match between them, so maybe "Should not contributeAndFinalize when result digest does not match calback data hash"

Reply:
Done [Update missing callback data test](https://github.com/iExecBlockchainComputing/PoCo/pull/176/commits/1b0b295e18402e58f4469e9099c527ea04bd6e10)